### PR TITLE
added user correlation query

### PR DESCRIPTION
**Pull request by habdelra**

### CODE REVIEW

File: packages/authentication/cardstack/middleware.js [hunk 1]
```diff
@@ -164,15 +166,23 @@ class Authentication {
     if (!user.data || !user.data.type) { return; }
 
     let have;
+    let query = source['userCorrelationQuery'];
 
-    if (user.data.id != null) {
-      try {
+    try {
+      if (user.data.id != null) {
         have = await this.userSearcher.get(user.data.type, user.data.id);
```
Comment (ef4) [inline packages/authentication/cardstack/middleware.js]: I don't think this should take priority over the userCorrelationQuery. It's legit to want to have externally -provided IDs while also doing custom correlation.

I think you can get rid of this step entirely by making the default userCorrelationQuery be one that just finds by ID.

File: packages/authentication/cardstack/middleware.js [hunk 2]
```diff
@@ -164,15 +166,23 @@ class Authentication {
...
+        let compiled = Handlebars.compile(query);
+        have = await this.userSearcher.search(JSON.parse(compiled(user)));
       }
+    } catch (err) {
+      if (err.status !== 404) {
+        throw err;
+      }
+    }
+
+    if (have && have.data && _.isArray(have.data)) {
```
Comment (ef4) [inline packages/authentication/cardstack/middleware.js]: We have ECMA standard `Array.isArray` now, no need for lodash here.

File: packages/authentication/cardstack/middleware.js [hunk 3]
```diff
@@ -164,15 +166,23 @@ class Authentication {
...
+      have = have.data.length ? _.merge({ data: have.data[0] }, user) : null; // deep merge
```
Comment (ef4) [inline packages/authentication/cardstack/middleware.js]: Why are we merging? I think the only time we would apply `user` to `have` is later if `mayUpdateUser`, and in that case the writers system does it for us.

In other words, I don't think anything about merging needs to change in this PR.

File: packages/authentication/cardstack/middleware.js (hunk 1 above)
Comment (habdelra) [reply packages/authentication/cardstack/middleware.js]: ah--good idea. will do

File: packages/authentication/cardstack/middleware.js (hunk 3 above)
Comment (habdelra) [reply packages/authentication/cardstack/middleware.js]: sure, no problem. I was actually think about this on my train ride in, thinking that we actually don't want auth0 to stomp on data here: like what if a user wants to change their name from how it appears in auth0. totally makes sense

**Pull request merged**